configuration/configmgr/configmgrlib/SchemaItem.cpp

```diff
@@ -652,88 +660,115 @@ void SchemaItem::setRequiredInstanceComponents(const std::string list)
+                    envNodesByParent[pParentEnvNode->getId()] = std::vector<std::shared_ptr<EnvironmentNode>>();
+                }
             }
         }
-    }
 
-    //
-    // Now build the vector per parent node ID
-    for (auto &pWeakEnvNode: m_envNodes)
-    {
-        std::shared_ptr<EnvironmentNode> pEnvNode = pWeakEnvNode.lock();
-        if (pEnvNode)
+        //
+        // Now build the vector per parent node ID
+        for (auto &pWeakEnvNode: m_envNodes)
         {
-            std::shared_ptr<EnvironmentNode> pParentEnvNode = pEnvNode->getParent();
-            if (pParentEnvNode)
+            std::shared_ptr<EnvironmentNode> pEnvNode = pWeakEnvNode.lock();
+            if (pEnvNode)
             {
-                envNodesByParent[pParentEnvNode->getId()].emplace_back(pEnvNode);
+                std::shared_ptr<EnvironmentNode> pParentEnvNode = pEnvNode->getParent();
+                if (pParentEnvNode)
+                {
+                    envNodesByParent[pParentEnvNode->getId()].emplace_back(pEnvNode);
+                }
             }
         }
-    }
 
-    //
-    // Now validate the count of nodes
-    for (auto &envNodes: envNodesByParent)
-    {
-        if (envNodes.second.size() < m_minInstances || envNodes.second.size() > m_maxInstances)
+        //
+        // Now validate the count of nodes
+        for (auto &envNodes: envNodesByParent)
         {
-            for (auto &pEnvNode: envNodes.second)
+            if (envNodes.second.size() < m_minInstances || envNodes.second.size() > m_maxInstances)
             {
-                std::string path;
-                getPath(path);
-                std::string msg;
-                msg = "The number of SchemaItem " + path + " nodes (" + std::to_string(envNodes.second.size()) +
-                      ") is outside the range of " + std::to_string(m_minInstances) +
-                      " to " + std::to_string(m_maxInstances);
-                status.addMsg(statusMsg::error, pEnvNode->getId(), "", msg);
+                for (auto &pEnvNode: envNodes.second)
+                {
+                    std::string path;
+                    getPath(path);
+                    std::string msg;
+                    msg = "The number of " + path + " nodes (" + std::to_string(envNodes.second.size()) +
+                          ") is outside the range of " + std::to_string(m_minInstances) +
+                          " to " + std::to_string(m_maxInstances);
+                    status.addMsg(statusMsg::error, pEnvNode->getId(), "", msg);
+                }
             }
         }
-    }
 
 
-    //
-    // Now validate each environment node (at this time go ahead and reset the member variable of weak pointers as well
-    // to get rid of any stale values)
-    for (auto &pWeakEnvNode: m_envNodes)
-    {
-        if (!isHidden() || includeHiddenNodes)
+        //
+        // Check the attributes for uniqueness.
+        for (auto &attribute: m_attributes)
```

Comment:
I don't know how many attributes we are searching here, but would we ever run into the case where the n^2 search here might get out of hand? Sorting  & using something like http://www.cplusplus.com/reference/algorithm/set_intersection/  might also be less code.

Reply:
I don't think this is an issue, however that could be an improvement. I shall do some research.